File: packages/ios/ByokyTests/TranslationLiveSmokeTest.swift
import XCTest
@testable import Byoky

/// Live smoke tests against real provider APIs on iOS.
///
/// Each test:
///   1. Uses TranslationEngine to translate a tiny anthropic-dialect request
///      into the target family's dialect.
///   2. Sends that translated body to the real provider via URLSession.
///   3. Translates the response back through the engine.
///   4. Asserts the round-trip parses cleanly.
///
/// Tests are gated behind environment variables. Skipped (XCTSkip) if a key
/// isn't set, so absence is non-fatal.
///
/// Run with all four families:
///   export BYOKY_TEST_ANTHROPIC_KEY=sk-ant-...
///   export BYOKY_TEST_OPENAI_KEY=sk-...
///   export BYOKY_TEST_GEMINI_KEY=AIza...
///   export BYOKY_TEST_COHERE_KEY=...
///   xcodebuild test -scheme Byoky -destination 'platform=iOS Simulator,name=iPhone 16'
///
/// Live tests are slow and rate-limited. Run sparingly.
final class TranslationLiveSmokeTest: XCTestCase {

    private let anthropicReqBody = """
    {"model":"claude-haiku-4-5-20251001","max_tokens":50,\
    "messages":[{"role":"user","content":"Reply with the single word: pong"}]}
    """

    func testAnthropicNativeRoundTrip() async throws {
        let key = try requireEnv("BYOKY_TEST_ANTHROPIC_KEY")
        let ctxJson = """
        {"srcFamily":"anthropic","dstFamily":"anthropic",\
        "srcProviderId":"anthropic","dstProviderId":"anthropic",\
        "srcModel":"claude-haiku-4-5-20251001","dstModel":"claude-haiku-4-5-20251001"}
        """

        let translated = try TranslationEngine.shared.translateRequest(
            contextJson: ctxJson,
            body: anthropicReqBody
        )
        let response = try await postJson(
            url: URL(string: "https://api.anthropic.com/v1/messages")!,
            body: translated,
            headers: [
                "x-api-key": key,
                "anthropic-version": "2023-06-01",
            ]
        )
        let parsed = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: Data(response.utf8)) as? [String: Any]
        )
        XCTAssertNotNil(parsed["content"], "anthropic response missing content[]")
    }

    func testOpenaiCrossFamilyRoundTrip() async throws {
        let key = try requireEnv("BYOKY_TEST_OPENAI_KEY")
        let ctxJson = """
        {"srcFamily":"anthropic","dstFamily":"openai",\
        "srcProviderId":"anthropic","dstProviderId":"openai",\
        "srcModel":"claude-haiku-4-5-20251001","dstModel":"gpt-4o-mini"}
        """

        let translatedReq = try TranslationEngine.shared.translateRequest(
            contextJson: ctxJson,
            body: anthropicReqBody
        )
        let raw = try await postJson(
            url: URL(string: "https://api.openai.com/v1/chat/completions")!,
            body: translatedReq,
            headers: ["Authorization": "Bearer \(key)"]
        )
        let translated = try TranslationEngine.shared.translateResponse(
            contextJson: ctxJson,
            body: raw
        )
        let parsed = try JSONSerialization.jsonObject(with: Data(translated.utf8)) as? [String: Any]
        XCTAssertNotNil(parsed?["content"], "translated response missing content[]")
    }

    func testGeminiCrossFamilyRoundTrip() async throws {
        let key = try requireEnv("BYOKY_TEST_GEMINI_KEY")
        let ctxJson = """
        {"srcFamily":"anthropic","dstFamily":"gemini",\
        "srcProviderId":"anthropic","dstProviderId":"gemini",\
        "srcModel":"claude-haiku-4-5-20251001","dstModel":"gemini-2.5-flash"}
        """

        let translatedReq = try TranslationEngine.shared.translateRequest(
            contextJson: ctxJson,
            body: anthropicReqBody
        )
        let raw = try await postJson(
            url: URL(string: "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent")!,
            body: translatedReq,
            headers: ["x-goog-api-key": key]
        )
        let translated = try TranslationEngine.shared.translateResponse(
            contextJson: ctxJson,
            body: raw
        )
        let parsed = try JSONSerialization.jsonObject(with: Data(translated.utf8)) as? [String: Any]
        XCTAssertNotNil(parsed?["content"], "translated response missing content[]")
    }

    func testCohereCrossFamilyRoundTrip() async throws {
        let key = try requireEnv("BYOKY_TEST_COHERE_KEY")
        let ctxJson = """
        {"srcFamily":"anthropic","dstFamily":"cohere",\
        "srcProviderId":"anthropic","dstProviderId":"cohere",\
        "srcModel":"claude-haiku-4-5-20251001","dstModel":"command-r-plus"}
        """

        let translatedReq = try TranslationEngine.shared.translateRequest(
            contextJson: ctxJson,
            body: anthropicReqBody
        )
        let raw = try await postJson(
            url: URL(string: "https://api.cohere.com/v2/chat")!,
            body: translatedReq,
            headers: ["Authorization": "Bearer \(key)"]
        )
        let translated = try TranslationEngine.shared.translateResponse(
            contextJson: ctxJson,
            body: raw
        )
        let parsed = try JSONSerialization.jsonObject(with: Data(translated.utf8)) as? [String: Any]
        XCTAssertNotNil(parsed?["content"], "translated response missing content[]")
    }

    // MARK: - Helpers

    private func requireEnv(_ name: String) throws -> String {
        guard let v = ProcessInfo.processInfo.environment[name], !v.isEmpty else {
            throw XCTSkip("\(name) not set")
        }
        return v
    }

    private func postJson(url: URL, body: String, headers: [String: String]) async throws -> String {
        var request = URLRequest(url: url, timeoutInterval: 60)
        request.httpMethod = "POST"
        request.httpBody = Data(body.utf8)
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        for (k, v) in headers { request.setValue(v, forHTTPHeaderField: k) }

        let (data, response) = try await URLSession.shared.data(for: request)
        let http = try XCTUnwrap(response as? HTTPURLResponse)
        let text = String(data: data, encoding: .utf8) ?? ""
        XCTAssertTrue(
            (200..<300).contains(http.statusCode),
            "upstream HTTP \(http.statusCode): \(text.prefix(200))"
        )
        return text
    }
}

File: packages/android/app/src/androidTest/java/com/byoky/app/proxy/TranslationLiveSmokeTest.kt
package com.byoky.app.proxy

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.byoky.app.BuildConfig
import okhttp3.OkHttpClient
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Assume.assumeTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.concurrent.TimeUnit

/**
 * Live smoke tests against real provider APIs.
 *
 * Each test:
 *   1. Asks TranslationEngine to translate a tiny anthropic-dialect request
 *      into the target family's dialect.
 *   2. Sends that translated body to the real provider via OkHttp.
 *   3. Translates the response back through the engine.
 *   4. Asserts the round-trip parses cleanly.
 *
 * This proves end-to-end correctness of the JS bridge against real network
 * conditions, without depending on Phase 2 (group routing) being landed yet.
 * The full proxy → bridge → upstream → bridge path is exercised at the
 * TranslationEngine layer, just bypassing ProxyService for now.
 *
 * Tests are gated behind environment-variable BuildConfig fields. If a key
 * isn't set, the test reports as skipped (assumeTrue) rather than failing.
 *
 * To run with all four families:
 *   export BYOKY_TEST_ANTHROPIC_KEY=sk-ant-...
 *   export BYOKY_TEST_OPENAI_KEY=sk-...
 *   export BYOKY_TEST_GEMINI_KEY=AIza...
 *   export BYOKY_TEST_COHERE_KEY=...
 *   ./gradlew connectedAndroidTest
 *
 * Live tests are slow and rate-limited. Run sparingly.
 */
@RunWith(AndroidJUnit4::class)
class TranslationLiveSmokeTest {
    private lateinit var engine: TranslationEngine
    private val http = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val anthropicReqBody = """
        {"model":"claude-haiku-4-5-20251001","max_tokens":50,
         "messages":[{"role":"user","content":"Reply with the single word: pong"}]}
    """.trimIndent().replace("\n", "")

    @Before
    fun setup() {
        val ctx = InstrumentationRegistry.getInstrumentation().targetContext
        engine = TranslationEngine.get(ctx)
        assumeTrue(
            "JavaScriptSandbox unsupported on this device",
            engine.isSupported
        )
    }

    @Test
    fun anthropicNativeRoundTrip() {
        val key = BuildConfig.TEST_ANTHROPIC_KEY
        assumeTrue("BYOKY_TEST_ANTHROPIC_KEY not set", key.isNotEmpty())

        // Same family — no translation needed, but we still go through the
        // engine surface to validate translateRequest is a no-op when src==dst.
        val ctxJson = """
            {"srcFamily":"anthropic","dstFamily":"anthropic",
             "srcProviderId":"anthropic","dstProviderId":"anthropic",
             "srcModel":"claude-haiku-4-5-20251001","dstModel":"claude-haiku-4-5-20251001"}
        """.trimIndent().replace("\n", "")

        val translated = engine.translateRequest(ctxJson, anthropicReqBody)
        val response = postJson(
            url = "https://api.anthropic.com/v1/messages",
            body = translated,
            headers = mapOf(
                "x-api-key" to key,
                "anthropic-version" to "2023-06-01",
            )
        )
        assertNotNull(response)
        val parsed = JSONObject(response!!)
        // Anthropic responses have content[].text
        assertTrue("anthropic response missing content[]", parsed.has("content"))
    }

    @Test
    fun openaiCrossFamilyRoundTrip() {
        val key = BuildConfig.TEST_OPENAI_KEY
        assumeTrue("BYOKY_TEST_OPENAI_KEY not set", key.isNotEmpty())

        val ctxJson = """
            {"srcFamily":"anthropic","dstFamily":"openai",
             "srcProviderId":"anthropic","dstProviderId":"openai",
             "srcModel":"claude-haiku-4-5-20251001","dstModel":"gpt-4o-mini"}
        """.trimIndent().replace("\n", "")

        val translatedReq = engine.translateRequest(ctxJson, anthropicReqBody)
        val rawResponse = postJson(
            url = "https://api.openai.com/v1/chat/completions",
            body = translatedReq,
            headers = mapOf("Authorization" to "Bearer $key"),
        )
        assertNotNull(rawResponse)

        // Translate openai response → anthropic dialect, then assert anthropic shape.
        val translatedResp = engine.translateResponse(ctxJson, rawResponse!!)
        val parsed = JSONObject(translatedResp)
        assertTrue("translated response missing content[]", parsed.has("content"))
    }

    @Test
    fun geminiCrossFamilyRoundTrip() {
        val key = BuildConfig.TEST_GEMINI_KEY
        assumeTrue("BYOKY_TEST_GEMINI_KEY not set", key.isNotEmpty())

        val ctxJson = """
            {"srcFamily":"anthropic","dstFamily":"gemini",
             "srcProviderId":"anthropic","dstProviderId":"gemini",
             "srcModel":"claude-haiku-4-5-20251001","dstModel":"gemini-2.5-flash"}
        """.trimIndent().replace("\n", "")

        val translatedReq = engine.translateRequest(ctxJson, anthropicReqBody)
        val rawResponse = postJson(
            url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent",
            body = translatedReq,
            headers = mapOf("x-goog-api-key" to key),
        )
        assertNotNull(rawResponse)

        val translatedResp = engine.translateResponse(ctxJson, rawResponse!!)
        val parsed = JSONObject(translatedResp)
        assertTrue("translated response missing content[]", parsed.has("content"))
    }

    @Test
    fun cohereCrossFamilyRoundTrip() {
        val key = BuildConfig.TEST_COHERE_KEY
        assumeTrue("BYOKY_TEST_COHERE_KEY not set", key.isNotEmpty())

        val ctxJson = """
            {"srcFamily":"anthropic","dstFamily":"cohere",
             "srcProviderId":"anthropic","dstProviderId":"cohere",
             "srcModel":"claude-haiku-4-5-20251001","dstModel":"command-r-plus"}
        """.trimIndent().replace("\n", "")

        val translatedReq = engine.translateRequest(ctxJson, anthropicReqBody)
        val rawResponse = postJson(
            url = "https://api.cohere.com/v2/chat",
            body = translatedReq,
            headers = mapOf("Authorization" to "Bearer $key"),
        )
        assertNotNull(rawResponse)

        val translatedResp = engine.translateResponse(ctxJson, rawResponse!!)
        val parsed = JSONObject(translatedResp)
        assertTrue("translated response missing content[]", parsed.has("content"))
    }

    private fun postJson(url: String, body: String, headers: Map<String, String>): String? {
        val req = Request.Builder()
            .url(url)
            .post(body.toRequestBody("application/json".toMediaTypeOrNull()))
            .apply {
                headers.forEach { (k, v) -> header(k, v) }
                header("content-type", "application/json")
            }
            .build()
        http.newCall(req).execute().use { resp ->
            val text = resp.body?.string() ?: return null
            assertTrue(
                "upstream HTTP ${resp.code}: ${text.take(200)}",
                resp.isSuccessful
            )
            return text
        }
    }
}
